h5py/tests/test_vds/test_highlevel_vds.py

```diff
@@ -398,5 +398,49 @@ def test_relative_vds(self):
     def tearDown(self):
         shutil.rmtree(self.tmpdir)
 
+@ut.skipUnless(vds_support,
+               'VDS requires HDF5 >= 1.9.233')
+class VDSUnlimitedTestCase(ut.TestCase):
+
+    def setUp(self):
+        self.tmpdir = tempfile.mkdtemp()
+        self.path = osp.join(self.tmpdir, "resize.h5")
+        with h5.File(self.path, "w") as f:
+            source_dset = f.create_dataset(
+                "source",
+                data=np.arange(20),
+                shape=(10, 2),
+                maxshape=(None, 2),
+                chunks=(10, 1),
+                fillvalue=-1
+            )
+            self.layout = h5.VirtualLayout((10, 1), np.int, (None, 1))
+            layout_source = h5.VirtualSource(source_dset)
+            self.layout[:h5.UNLIMITED, 0] = layout_source[:h5.UNLIMITED, 1]
+
+            virtual_dset = f.create_virtual_dataset("virtual", self.layout)
+
+    def test_unlimited_axis(self):
+        comp1 = np.arange(1, 20, 2).reshape(10, 1)
+        comp2 = np.vstack((
+            comp1,
+            np.full(shape=(10, 1), fill_value=-1)
+        ))
+        comp3 = np.vstack((
+            comp1,
+            np.full(shape=(10, 1), fill_value=0)
+        ))
+        with h5.File(self.path, "a") as f:
+            source_dset = f['source']
+            virtual_dset = f['virtual']
+            assert (comp1 == virtual_dset).all()
+            source_dset.resize(20, 0)
+            assert (comp2 == virtual_dset).all()
```

Comment:
I'm confused - you're now checking against a `(20, 1)` array, but the virtual dataset should still have shape `(10, 1)`. Surely it doesn't automatically resize when the source dataset resizes?

Reply:
It surely does, no manual resizing required. If this is a bug, its an awfully convenient one.

Reply:
OK! I probably wouldn't design it like that, but I can't see how it could be a bug in h5py - we get the shape from HDF5.

Does that mean that a virtual dataset in a read-only file can resize itself with no warning? We currently assume that it's safe to cache a dataset's shape if it's in a read-only file.

Reply:
A read-only opened file would never have anything change unless opened in SWMR. So VDS shouldnt affect anything in that way by itself, but your caching decision should be determined off of `writable or swmr` not just `writable`. If that case is true (haven't looked myself), then the cached value shouldn't ever get invalidated.

Reply:
But if a virtual dataset can change shape based on its source data changing, and its source data is in a different file, you could have the virtual dataset open in a read-only file and the source dataset in a read-write file. Or have I missed something?

With SWMR, as far as I know you have to call `.refresh()` to update the dataset metadata, so we're using that to invalidate cached information at the h5py level as well.

Reply:
If the virtual source is in another file, that file would have to be opened with the same permissions/mode as the original file the virtual dataset is in. So if the original is in write mode, linked dataset is opened in write, etc. Therefore even if the dataset was in a different file, it would have the same constraints.

Reply:
I tweaked your test to illustrate what I mean. It is a problem, but this PR doesn't introduce it, it just makes it easier to see. Luckily, there's an easy workaround: call `.refresh()` to clear the cached shape.

```python
import h5py
import numpy as np
import os.path as osp
import tempfile

tmpdir = tempfile.mkdtemp()
path = osp.join(tmpdir, "resize.h5")
vds_path = osp.join(tmpdir, "resize_vds.h5")
with h5py.File(path, "w") as f:
    source_dset = f.create_dataset(
        "source",
        data=np.arange(20),
        shape=(10, 2),
        maxshape=(None, 2),
        chunks=(10, 1),
        fillvalue=-1
    )
    layout = h5py.VirtualLayout((10, 1), np.int, maxshape=(None, 1))
    layout_source = h5py.VirtualSource(source_dset)
    layout[:h5py.UNLIMITED, 0] = layout_source[:h5py.UNLIMITED, 1]

with h5py.File(vds_path, "w") as f:
    f.create_virtual_dataset("virtual", layout)


with h5py.File(path, "a") as f, h5py.File(vds_path, 'r') as fv:
    source_dset = f['source']
    virtual_dset = fv['virtual']

    print("Before", virtual_dset.shape)  # (10, 1)
    source_dset.resize(20, axis=0)
    print("After resize", virtual_dset.shape)  # (10, 1)
    virtual_dset.refresh()
    print("After refresh", virtual_dset.shape)  # (20, 1)
```

Reply:
I've opened #1693 about this.